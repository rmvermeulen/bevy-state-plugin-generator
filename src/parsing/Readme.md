# parsing

Parsing is implemented using [`nom`](crates.io/nom). The result of a successfully
parsed text is the root `Node` which contains any children.

See the [tests](./tests.rs) for detailed examples of parsing behavior.
