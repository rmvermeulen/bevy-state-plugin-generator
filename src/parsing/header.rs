use itertools::Itertools;
use lazy_regex::regex;

use crate::generate::get_package_info;
use crate::{NamingScheme, PluginConfig};

#[derive(Clone, Debug, Default, PartialEq)]
pub(crate) struct TemplateHeader<'a> {
    pub template: Vec<&'a str>,
    pub comments_block: Vec<&'a str>,
    pub info_block: Vec<String>,
}

trait Unquote {
    fn unquoted(&self) -> Self;
}

impl Unquote for &str {
    fn unquoted(&self) -> Self {
        self.strip_prefix("\"")
            .and_then(|s| s.strip_suffix("\""))
            .unwrap_or(self)
    }
}

pub(crate) fn parse_template_header<'a>(
    source: &'a str,
    plugin_config: &mut PluginConfig,
) -> TemplateHeader<'a> {
    let leading_comments = source
        .lines()
        .take_while(|line| line.starts_with("//"))
        .collect_vec();

    let mut info_block = Vec::new();
    info_block.push(format!("generated by {}", get_package_info()));

    macro_rules! warn {
        ($($exprs:expr),*) => {
            let warning = format!($($exprs),*);
            let warning = format!("WARN: {warning}");
            eprintln!("{warning}");
            info_block.push(warning);
        };
    }

    let mut template_source = Vec::new();
    let mut in_template = false;
    let leading_comments = leading_comments
        .iter()
        .filter(|line| {
            if in_template {
                if let Some(line) = line.strip_prefix("//") {
                    template_source.push(line.trim());
                    true
                } else {
                    false
                }
            } else if let Some(captures) =
                regex!(r#"^\s*//\s*bspg:(\w+)\s+(\w+|"\w+\")\s*$"#).captures(line)
            {
                let (_, [name, value]) = captures.extract();
                match name {
                    "root_state_name" => {
                        plugin_config.root_state_name = if value == "None" {
                            None
                        } else {
                            Some(value.unquoted().to_string())
                        };
                    }
                    "naming_scheme" => {
                        if let Some(naming_scheme) = NamingScheme::try_parse(value) {
                            plugin_config.naming_scheme = naming_scheme;
                        } else {
                            warn!("invalid naming scheme '{value}' (expected [none, short, full])");
                        }
                    }
                    _ => {
                        warn!("unknown setting: '{name}'");
                    }
                }
                true
            } else if regex!(r#"^\s*//\s*bspg:\s*$"#).is_match(line) {
                in_template = true;
                true
            } else {
                eprintln!("dropping: '{line:?}'");
                false
            }
        })
        .copied()
        .collect_vec();
    TemplateHeader {
        template: template_source,
        comments_block: leading_comments,
        info_block,
    }
}

#[cfg(test)]
mod tests {
    use bevy_utils::default;
    use indoc::formatdoc;
    use rstest::rstest;
    use speculoos::assert_that;
    use speculoos::prelude::{ContainingIntoIterAssertions, VecAssertions};

    use crate::parsing::header::parse_template_header;
    use crate::{NamingScheme, PluginConfig};

    #[rstest]
    #[case(String::new())]
    #[case::ignore_generated_lines(formatdoc! {"
        // generated by v[CARGO_PKG_VERSION]
        // WARN: some warning
    "})]
    fn test_parse_template_header(#[case] header: String) {
        let mut config: PluginConfig = default();
        let header = parse_template_header(&header, &mut config);
        assert_that!(header.template).is_empty();
        assert_that!(header.comments_block).is_empty();
    }

    #[rstest]
    fn test_parse_template_header_keep_variables() {
        let header = formatdoc! {"
            // generated by v[CARGO_PKG_VERSION]
            // WARN: some warning
            // bspg:some setting
        "};
        let mut config: PluginConfig = default();
        let header = parse_template_header(&header, &mut config);
        assert_that!(header.comments_block).contains("// bspg:some setting");
        assert_that!(header.info_block).contains("WARN: unknown setting: 'some'".to_string());
    }

    #[rstest]
    fn test_parse_template_header_variable_values(
        #[values(NamingScheme::Full, NamingScheme::None, NamingScheme::Short)]
        naming_scheme: NamingScheme,
        #[values(None, Some("MyStateRoot"))] name: Option<&str>,
    ) {
        let mut config: PluginConfig = default();
        let header = formatdoc! {"
            // generated by v[CARGO_PKG_VERSION]
            // bspg:root_state_name {}
            // bspg:naming_scheme   {}
            // bspg:
            ",
            name.map(|name| format!("\"{name}\"")).unwrap_or("None".to_string()),
            naming_scheme
        };
        let _ = parse_template_header(&header, &mut config);
        assert_that!(config.root_state_name).is_equal_to(name.map(String::from));
        assert_that!(config.naming_scheme).is_equal_to(naming_scheme);
    }
}
