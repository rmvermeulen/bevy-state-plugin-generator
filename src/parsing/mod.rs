pub(crate) mod header;
mod parsers;
#[cfg(test)]
mod tests;
mod tokens;

pub use parsers::*;
pub(crate) use tokens::*;

// TODO: investigate "tree-less parsing", see https://www.youtube.com/watch?v=NxiKlnUtyio
