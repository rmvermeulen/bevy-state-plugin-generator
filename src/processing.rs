#[cfg(test)]
mod tests;

use std::collections::VecDeque;

use bevy_platform::collections::HashSet;
use bevy_utils::default;
use indoc::formatdoc;
use itertools::{Itertools, concat};

use crate::generate::REQUIRED_DERIVES;
use crate::parsing::ParseNode;
use crate::{NamingScheme, PluginConfig, PluginName};

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub enum NodeType {
    #[default]
    Singleton,
    List,
    Enum,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct NodeData {
    pub index: usize,
    pub parent: Option<usize>,
    pub node_type: NodeType,
    pub depth: usize,
    pub name: String,
    pub resolved_name: Option<String>,
    pub variants: Vec<String>,
}

pub fn flatten_root_parse_node(root_node: ParseNode<'_>) -> Vec<NodeData> {
    let node_count = root_node.get_tree_size();
    let mut nodes = Vec::with_capacity(node_count);
    let mut todo = VecDeque::from([(root_node, 0, None)]);
    while let Some((parse_node, depth, parent)) = todo.pop_front() {
        let node_type = match parse_node {
            ParseNode::Singleton(_) => NodeType::Singleton,
            ParseNode::Enum(_, _) => NodeType::Enum,
            ParseNode::List(_, _) => NodeType::List,
            ParseNode::Comment(_) => {
                continue;
            }
        };
        let Some(name) = parse_node.name() else {
            continue;
        };
        let index = nodes.len();
        nodes.push(NodeData {
            node_type,
            index,
            parent,
            depth,
            name: name.to_string(),
            ..default()
        });
        for child in parse_node.children() {
            todo.push_back((child, depth + 1, Some(index)));
        }
    }

    for i in 0..nodes.len() {
        let (ancestors, children) = nodes.split_at_mut(i);
        let node = &children[0];
        if let Some(parent_index) = node.parent {
            assert!(parent_index < i);
            let parent = &mut ancestors[parent_index];
            parent.variants.push(node.name.clone());
        }
    }

    #[cfg(debug_assertions)]
    for (i, node) in nodes.iter().enumerate() {
        assert_eq!(node.index, i);
        if let Some(parent_index) = node.parent {
            assert!(parent_index <= node.index);
        }
    }

    nodes
}

pub fn apply_naming_scheme(
    naming_scheme: NamingScheme,
    nodes: &mut [NodeData],
) -> Result<(), ProcessingError> {
    let mut names = HashSet::new();
    let mut resolved_names = Vec::new();
    for node in nodes.iter() {
        let base_name = &node.name;
        let resolved_name = match naming_scheme {
            NamingScheme::Full => {
                let mut ancestors = Vec::new();
                let mut current = node;
                while let Some(p) = current.parent {
                    current = &nodes[p];
                    ancestors.push(current);
                }
                let ancestral_name = ancestors.into_iter().map(|a| &a.name).rev().join("");
                format!("{ancestral_name}{base_name}")
            }
            NamingScheme::Short => {
                let parent = node
                    .parent
                    .map(|p| nodes[p].name.as_str())
                    .unwrap_or_default();
                format!("{parent}{base_name}")
            }
            NamingScheme::None => base_name.clone(),
        };
        if !names.insert(resolved_name.clone()) {
            return Err(ProcessingError::DuplicateName {
                resolved_name,
                original_name: base_name.clone(),
            });
        }

        resolved_names.push(resolved_name);
    }
    assert_eq!(names.len(), resolved_names.len());
    for (i, resolved_name) in resolved_names.into_iter().enumerate() {
        nodes[i].resolved_name = Some(resolved_name);
    }
    Ok(())
}

type NomError<T> = nom::Err<nom::error::Error<T>>;

#[derive(Debug, thiserror::Error)]
pub enum ProcessingError {
    #[error("Duplicate name: resolved_name='{resolved_name}' original_name='{original_name}'")]
    DuplicateName {
        resolved_name: String,
        original_name: String,
    },
    #[error("Unspecified error: {0}")]
    Custom(String),
    #[error("Failed to parse! Final state: {0:?}")]
    Parsing(#[from] nom::Err<nom::error::Error<String>>),
    #[error("Invalid configuration: {0}")]
    InvalidConfig(String),
}

impl From<NomError<&str>> for ProcessingError {
    fn from(value: NomError<&str>) -> Self {
        Self::from(value.to_owned())
    }
}

fn build_plugin_source(
    nodes: Vec<NodeData>,
    config: PluginConfig,
) -> Result<String, ProcessingError> {
    let PluginConfig {
        additional_derives: derives,
        plugin_name,
        root_state_name,
        naming_scheme: _,
        states_module_name,
    } = config;

    let derives = concat([
        REQUIRED_DERIVES
            .iter()
            .map(ToString::to_string)
            .collect_vec(),
        derives.into_iter().collect_vec(),
    ])
    .into_iter()
    .unique()
    .join(", ");
    let definitions_source = nodes
        .iter()
        .map(|node| {
            let derives = node
                .parent
                .map(|parent_id| {
                    let parent = &nodes[parent_id];
                    formatdoc! {"
                        #[derive(bevy::prelude::SubStates, {derives})]
                        #[source({source} = {source}::{variant})]
                    ", variant = node.name, source = parent.name
                    }
                })
                .unwrap_or_else(|| formatdoc! {"#[derive(bevy::prelude::States, {derives})]"})
                .trim()
                .to_string();
            let typename = &node.name;

            let source_for_singleton = || {
                formatdoc! {"
                {derives}
                pub struct {typename};
            "}
            };
            let source_for_enum = |variants: &[&str]| {
                formatdoc! {"
                {derives}
                pub enum {typename} {{
                    #[default]
                    {variants}
                }}
                ", variants = variants.join(",\n      ")
                }
            };

            match node.node_type {
                NodeType::Singleton => source_for_singleton(),
                NodeType::Enum => {
                    if node.variants.is_empty() {
                        source_for_singleton()
                    } else {
                        let variants = node.variants.iter().map(String::as_str).collect_vec();
                        source_for_enum(&variants)
                    }
                }
                NodeType::List => source_for_singleton(),
            }
        })
        .join("\n");

    let plugin_builder = if let Some(root_state_name) = root_state_name.as_ref() {
        let states_module_name = states_module_name.as_str();
        let init_state = format!(".init_state::<{states_module_name}::{root_state_name}>()");
        let sub_states = nodes
            .iter()
            .flat_map(|node| {
                let resolved_name = node
                    .resolved_name
                    .as_ref()
                    .expect("Node name has not been resolved!");
                if node.parent.map(|p| nodes[p].clone()).is_some() {
                    Some(format!(
                        ".add_sub_state::<{states_module_name}::{resolved_name}>()"
                    ))
                } else {
                    None
                }
            })
            .join("\n            ");
        format!("app{init_state}{sub_states};")
    } else {
        let states_module_name = states_module_name.as_str();
        let states = nodes
            .iter()
            .map(|node| {
                let resolved_name = node
                    .resolved_name
                    .clone()
                    .expect("Node name has not been resolved!");
                if node.parent.map(|p| nodes[p].clone()).is_some() {
                    format!(".add_sub_state::<{states_module_name}::{resolved_name}>()")
                } else {
                    format!(".add_state::<{states_module_name}::{resolved_name}>()")
                }
            })
            .join("\n            ");
        format!("app{states};")
    };

    let plugin_def = match plugin_name {
        PluginName::Struct(plugin_name) => {
            formatdoc! {"
                pub struct {plugin_name};
                impl bevy::app::Plugin for {plugin_name} {{
                    fn build(&self, app: &mut bevy::app::App) {{
                        {plugin_builder}
                    }}
                }}
            "}
        }
        PluginName::Function(plugin_name) => {
            formatdoc! {"
                pub fn {plugin_name}(app: &mut bevy::app::App) {{
                    {plugin_builder}
                }}
            "}
        }
    };

    Ok(formatdoc! {"
        use bevy::prelude::AppExtStates;
        #[allow(missing_docs)]
        pub mod {states_module_name} {{
            use bevy::prelude::StateSet;
            {definitions_source}
        }}
        {plugin_def}
    "})
}

pub(crate) fn remove_root_node(nodes: &mut Vec<NodeData>) {
    nodes.remove(0);
    for node in nodes {
        assert!(node.index > 0);
        node.index -= 1;
        assert!(node.depth > 0);
        node.depth -= 1;
        if let Some(parent_index) = &mut node.parent {
            node.parent = (*parent_index > 0).then(|| *parent_index - 1);
        }
    }
}

pub(crate) fn process_parse_nodes(
    parse_nodes: Vec<ParseNode<'_>>,
    naming_scheme: NamingScheme,
    root_state_name: Option<String>,
) -> Result<Vec<NodeData>, ProcessingError> {
    // add the implicit root_node according to config
    let root_node = if let Some(root_state_name) = &root_state_name {
        ParseNode::Enum(root_state_name.as_str().into(), parse_nodes)
    } else {
        // add a temporary root node
        ParseNode::List("".into(), parse_nodes)
    };
    let mut nodes = flatten_root_parse_node(root_node);
    // remove temporary root node
    if root_state_name.is_none() {
        remove_root_node(&mut nodes);
    }
    apply_naming_scheme(naming_scheme, &mut nodes)?;
    Ok(nodes)
}

pub fn convert_parse_nodes_into_plugin_source(
    parse_nodes: Vec<ParseNode<'_>>,
    config: PluginConfig,
) -> Result<String, ProcessingError> {
    let nodes = process_parse_nodes(
        parse_nodes,
        config.naming_scheme,
        config.root_state_name.clone(),
    )?;
    assert!(nodes.iter().all(|node| node.resolved_name.is_some()));
    build_plugin_source(nodes, config)
}
