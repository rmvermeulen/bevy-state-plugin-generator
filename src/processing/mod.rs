#[cfg(test)]
mod tests;

use std::collections::VecDeque;

use bevy_platform::collections::HashSet;
use bevy_utils::default;
use indoc::formatdoc;
use itertools::{Itertools, concat};

use crate::generate::REQUIRED_DERIVES;
use crate::parsing::ParseNode;
use crate::tree::SubTree;
use crate::{NamingScheme, PluginConfig, PluginName};

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub enum NodeType {
    #[default]
    Singleton,
    List,
    Enum,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct NodeData {
    pub index: usize,
    pub parent: Option<usize>,
    pub node_type: NodeType,
    pub depth: usize,
    pub name: String,
    pub resolved_name: Option<String>,
    pub variants: Vec<String>,
}

pub fn flatten_parse_node(root_node: ParseNode<'_>) -> Vec<NodeData> {
    let node_count = root_node.get_tree_size();
    let mut nodes = Vec::with_capacity(node_count);
    let mut todo = VecDeque::from([(root_node, 0, None)]);
    while let Some((parse_node, depth, parent)) = todo.pop_front() {
        let Some(name) = parse_node.name() else {
            continue;
        };
        let node_type = match parse_node {
            ParseNode::Singleton(_) => NodeType::Singleton,
            ParseNode::Enum(_, _) => NodeType::Enum,
            ParseNode::List(_, _) => NodeType::List,
            ParseNode::Comment(_) => {
                unreachable!("Comment has no name")
            }
        };
        let index = nodes.len();
        nodes.push(NodeData {
            node_type,
            index,
            parent,
            depth,
            name: name.to_string(),
            ..default()
        });
        for child in parse_node.children() {
            todo.push_back((child, depth + 1, Some(index)));
        }
    }

    for (i, node) in nodes.iter().enumerate() {
        assert_eq!(node.index, i);
        if let Some(parent) = node.parent {
            assert!(parent <= node.index);
            assert!(node.depth > nodes[parent].depth);
        }
    }

    nodes
}

pub fn apply_naming_scheme(
    naming_scheme: NamingScheme,
    nodes: &mut [NodeData],
) -> Result<(), ProcessingError> {
    let mut names = HashSet::new();
    let mut resolved_names = Vec::new();
    for node in nodes.iter() {
        let base_name = &node.name;
        let resolved_name = match naming_scheme {
            NamingScheme::Full => {
                let mut ancestors = Vec::new();
                let mut current = node;
                while let Some(p) = current.parent {
                    current = &nodes[p];
                    ancestors.push(current);
                }
                let ancestral_name = ancestors.into_iter().map(|a| &a.name).join("");
                format!("{ancestral_name}{base_name}")
            }
            NamingScheme::Short => {
                let parent = node
                    .parent
                    .map(|p| nodes[p].name.as_str())
                    .unwrap_or_default();
                format!("{parent}{base_name}")
            }
            NamingScheme::None => base_name.clone(),
        };
        if !names.insert(resolved_name.clone()) {
            return Err(ProcessingError::DuplicateName {
                resolved_name,
                original_name: base_name.clone(),
            });
        }

        resolved_names.push(resolved_name);
    }
    assert_eq!(names.len(), resolved_names.len());
    for (i, resolved_name) in resolved_names.into_iter().enumerate() {
        nodes[i].resolved_name = Some(resolved_name);
    }
    Ok(())
}

#[derive(Debug, thiserror::Error)]
pub enum ProcessingError {
    #[error("Duplicate name: resolved_name='{resolved_name}' original_name='{original_name}'")]
    DuplicateName {
        resolved_name: String,
        original_name: String,
    },
    #[error("Unspecified error: {0}")]
    Custom(String),
    #[error("Failed to parse! Final state: {0:?}")]
    Parsing(String),
}

fn get_source(nodes: Vec<NodeData>, config: PluginConfig) -> Result<String, ProcessingError> {
    let PluginConfig {
        additional_derives: derives,
        plugin_name,
        root_state_name,
        naming_scheme: _,
        states_module_name,
    } = config;

    // add the implicit root_node according to config
    let nodes = if let Some(root_state_name) = &root_state_name {
        let root_node = NodeData {
            index: 0,
            parent: None,
            name: root_state_name.clone(),
            resolved_name: Some(root_state_name.clone()),
            ..default()
        };
        let mut new_nodes = vec![root_node];
        for mut node in nodes.into_iter() {
            let root_node = &mut new_nodes[0];
            if node.depth == 0 {
                root_node.variants.push(
                    node.resolved_name
                        .clone()
                        .unwrap_or_else(|| panic!("Node name has not been resolved! {node:?}")),
                );
            }
            // adjust indices, since we've added a root_node
            node.index += 1;
            node.depth += 1;
            if let Some(parent_index) = &mut node.parent {
                *parent_index += 1;
            } else {
                // parent this node to the root_node
                _ = node.parent.insert(0);
            }
            assert!(node.parent.is_some());
            new_nodes.push(node);
        }
        new_nodes
    } else {
        nodes
    };

    let derives = concat([
        REQUIRED_DERIVES
            .iter()
            .map(ToString::to_string)
            .collect_vec(),
        derives.into_iter().collect_vec(),
    ])
    .into_iter()
    .unique()
    .join(", ");
    let definitions_source = nodes
        .iter()
        .map(|node| {
            let derives = node
                .parent
                .map(|parent_id| {
                    let parent = &nodes[parent_id];
                    formatdoc! {"
                        #[derive(bevy::prelude::SubStates, {derives})]
                        #[source({source} = {source}::{variant})]
                    ", variant = node.name, source = parent.name
                    }
                })
                .unwrap_or_else(|| formatdoc! {"#[derive(bevy::prelude::States, {derives})]"})
                .trim()
                .to_string();
            let typename = &node.name;

            let source_for_singleton = || {
                formatdoc! {"
                {derives}
                pub struct {typename};
            "}
            };
            let source_for_enum = |variants: &[&str]| {
                formatdoc! {"
                {derives}
                pub enum {typename} {{
                    #[default]
                    {variants}
                }}
                ", variants = variants.join(",\n      ")
                }
            };

            match node.node_type {
                NodeType::Singleton => source_for_singleton(),
                NodeType::Enum => {
                    if node.variants.is_empty() {
                        source_for_singleton()
                    } else {
                        let variants = node.variants.iter().map(String::as_str).collect_vec();
                        source_for_enum(&variants)
                    }
                }
                NodeType::List => source_for_singleton(),
            }
        })
        .join("\n");

    let plugin_builder = if let Some(root_state_name) = root_state_name.as_ref() {
        let states_module_name = states_module_name.as_str();
        let init_state = format!(".init_state::<{states_module_name}::{root_state_name}>()");
        let sub_states = nodes
            .iter()
            .flat_map(|node| {
                let resolved_name = node
                    .resolved_name
                    .as_ref()
                    .expect("Node name has not been resolved!");
                if node.parent.map(|p| nodes[p].clone()).is_some() {
                    Some(format!(
                        ".add_sub_state::<{states_module_name}::{resolved_name}>()"
                    ))
                } else {
                    None
                }
            })
            .join("\n            ");
        format!("app{init_state}{sub_states};")
    } else {
        let states_module_name = states_module_name.as_str();
        let states = nodes
            .iter()
            .map(|node| {
                let resolved_name = node
                    .resolved_name
                    .clone()
                    .expect("Node name has not been resolved!");
                if node.parent.map(|p| nodes[p].clone()).is_some() {
                    format!(".add_sub_state::<{states_module_name}::{resolved_name}>()")
                } else {
                    format!(".add_state::<{states_module_name}::{resolved_name}>()")
                }
            })
            .join("\n            ");
        format!("app{states};")
    };

    let plugin_def = match plugin_name {
        PluginName::Struct(plugin_name) => {
            formatdoc! {"
                pub struct {plugin_name};
                impl bevy::app::Plugin for {plugin_name} {{
                    fn build(&self, app: &mut bevy::app::App) {{
                        {plugin_builder}
                    }}
                }}
            "}
        }
        PluginName::Function(plugin_name) => {
            formatdoc! {"
                pub fn {plugin_name}(app: &mut bevy::app::App) {{
                    {plugin_builder}
                }}
            "}
        }
    };

    Ok(formatdoc! {"
        #![allow(missing_docs)]
        use bevy::prelude::AppExtStates;
        pub mod {states_module_name} {{
            use bevy::prelude::StateSet;
            {definitions_source}
        }}
        {plugin_def}
    "})
}

pub fn parse_node_into_final_source(
    node: ParseNode<'_>,
    config: PluginConfig,
) -> Result<String, ProcessingError> {
    let mut nodes = flatten_parse_node(node);
    apply_naming_scheme(config.naming_scheme, &mut nodes)?;
    assert!(nodes.iter().all(|node| node.resolved_name.is_some()));
    get_source(nodes, config)
}
