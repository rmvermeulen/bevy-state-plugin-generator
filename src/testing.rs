pub use insta::*;
pub use rstest::*;
pub use speculoos::prelude::*;

pub use crate::set_snapshot_suffix;

#[macro_export]
#[allow(missing_docs)]
macro_rules! set_snapshot_suffix {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(format!($($expr,)*));
        let _guard = settings.bind_to_scope();
    }
}

pub mod node_data {
    use bevy_utils::default;

    use super::*;
    use crate::processing::NodeData;

    fn node<S: ToString>(name: S, index: usize, depth: usize, parent: Option<usize>) -> NodeData {
        let name = name.to_string();
        NodeData {
            index,
            parent,
            resolved_name: Some(name.clone()),
            name,
            depth,
            ..default()
        }
    }

    #[fixture]
    pub fn single_node() -> Vec<NodeData> {
        vec![node("RootState", 0, 0, None)]
    }

    #[fixture]
    pub fn nested_example() -> Vec<NodeData> {
        vec![
            node("RootState", 0, 0, None),
            node("Menu", 1, 1, Some(0)),
            node("Options", 2, 2, Some(1)),
            node("Audio", 3, 3, Some(2)),
            node("Video", 4, 3, Some(2)),
        ]
    }
}

pub mod parse_node {
    use super::*;
    use crate::parsing::Node;

    #[fixture]
    pub fn duplicate_names() -> Node<'static> {
        Node::list("Root", [Node::singleton("A"), Node::singleton("A")])
    }

    #[fixture]
    pub fn overlapping_names() -> Node<'static> {
        Node::list(
            "Root",
            [
                Node::enumeration("Root", [Node::singleton("A")]),
                Node::singleton("A"),
            ],
        )
    }

    #[fixture]
    pub fn enum_root_a() -> Node<'static> {
        Node::enumeration("Root", [Node::singleton("A")])
    }

    #[fixture]
    pub fn enum_root_ab() -> Node<'static> {
        Node::enumeration("Root", [Node::singleton("A"), Node::singleton("B")])
    }

    #[fixture]
    pub fn enum_root_a_b() -> Node<'static> {
        Node::enumeration("Root", [Node::enumeration("A", [Node::singleton("B")])])
    }

    #[fixture]
    pub fn enum_root_a_b_up_c() -> Node<'static> {
        Node::enumeration(
            "Root",
            [
                Node::enumeration("A", [Node::singleton("B")]),
                Node::singleton("C"),
            ],
        )
    }

    #[fixture]
    pub fn list_root_a() -> Node<'static> {
        Node::list("Root", [Node::singleton("A")])
    }

    #[fixture]
    pub fn list_root_ab() -> Node<'static> {
        Node::list("Root", [Node::singleton("A"), Node::singleton("B")])
    }

    #[fixture]
    pub fn list_root_a_b() -> Node<'static> {
        Node::list("Root", [Node::list("A", [Node::singleton("B")])])
    }

    #[fixture]
    pub fn list_root_a_b_up_c() -> Node<'static> {
        Node::list(
            "Root",
            [
                Node::list("A", [Node::singleton("B")]),
                Node::singleton("C"),
            ],
        )
    }

    #[fixture]
    pub fn nested_example() -> Node<'static> {
        Node::enumeration(
            "Menu",
            [
                Node::singleton("Main"),
                Node::comment("these are the options"),
                Node::enumeration(
                    "Options",
                    [
                        Node::singleton("Graphics"),
                        Node::singleton("Audio"),
                        Node::singleton("Gameplay"),
                    ],
                ),
                Node::enumeration("Game", [Node::singleton("Save"), Node::singleton("Load")]),
            ],
        )
    }
}
