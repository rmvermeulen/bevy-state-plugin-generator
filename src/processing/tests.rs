use bevy_utils::default;
use itertools::Itertools;

use crate::NamingScheme;
use crate::parsing::ParseNode;
use crate::processing::{NodeData, NodeType, apply_naming_scheme, flatten_root_parse_node};
use crate::testing::*;

#[rstest]
fn test_flatten_parse_node_singleton() {
    let results = flatten_root_parse_node(ParseNode::singleton("A"));
    assert_that!(results).is_equal_to(vec![NodeData {
        index: 0,
        parent: None,
        node_type: NodeType::Singleton,
        depth: 0,
        name: "A".to_string(),
        ..default()
    }]);
}

#[rstest]
#[case::enum_root_a(parse_node::enum_root_a())]
#[case::enum_root_ab(parse_node::enum_root_ab())]
#[case::enum_root_a_b(parse_node::enum_root_a_b())]
#[case::enum_root_a_b_up_c(parse_node::enum_root_a_b_up_c())]
fn test_flatten_parse_node_enums(#[context] context: Context, #[case] node: ParseNode) {
    set_snapshot_suffix!("{}", context.description.unwrap());
    assert_debug_snapshot!(flatten_root_parse_node(node));
}

#[rstest]
#[case::list_root_a(parse_node::list_root_a())]
#[case::list_root_ab(parse_node::list_root_ab())]
#[case::list_root_a_b(parse_node::list_root_a_b())]
#[case::list_root_a_b_up_c(parse_node::list_root_a_b_up_c())]
fn test_flatten_parse_node_lists(#[context] context: Context, #[case] node: ParseNode) {
    set_snapshot_suffix!("{}", context.description.unwrap());
    assert_debug_snapshot!(flatten_root_parse_node(node));
}

#[rstest]
#[case::single_node(node_data::single_node())]
#[case::nested_example(node_data::nested_example())]
fn test_apply_naming_scheme(
    #[context] context: Context,
    #[case] mut nodes: Vec<NodeData>,
    #[values(NamingScheme::Short, NamingScheme::Full, NamingScheme::None)]
    naming_scheme: NamingScheme,
) {
    set_snapshot_suffix!("{}_{naming_scheme}", context.description.unwrap());
    apply_naming_scheme(naming_scheme, &mut nodes).unwrap();
    assert_debug_snapshot!(
        nodes
            .into_iter()
            .map(|node| format!("{} -> {}", node.name, node.resolved_name.unwrap()))
            .collect_vec()
    );
}

#[rstest]
fn test_apply_naming_scheme_differences(
    #[from(node_data::nested_example)] mut nodes: Vec<NodeData>,
) {
    let outputs = [NamingScheme::Short, NamingScheme::Full, NamingScheme::None]
        .into_iter()
        .map(|naming_scheme| {
            apply_naming_scheme(naming_scheme, &mut nodes).unwrap();
            nodes
                .iter()
                .map(|node| format!("{} -> {}", node.name, node.resolved_name.clone().unwrap()))
                .collect_vec()
        })
        .collect_vec();
    assert_that!(outputs).has_length(3);
    assert_debug_snapshot!(outputs);
}
