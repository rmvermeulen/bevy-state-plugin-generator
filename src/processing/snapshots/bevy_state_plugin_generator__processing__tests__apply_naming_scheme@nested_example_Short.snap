---
source: src/processing/tests.rs
expression: "nodes.into_iter().map(|node|\nformat!(\"{} -> {}\", node.name, node.resolved_name.unwrap())).collect_vec()"
---
[
    "RootState -> RootState",
    "Menu -> RootStateMenu",
    "Options -> MenuOptions",
    "Audio -> OptionsAudio",
    "Video -> OptionsVideo",
]
